feat: builds.output_path

tmeckel opened the pull request:
<!--

Hi, thanks for contributing!

Please make sure you read our CONTRIBUTING guide.

Also, add tests and the respective documentation changes as well.

-->


<!-- If applied, this commit will... -->

This Pull Request will add the option `output_path` to a build definition in the `config.yaml` file.

<!-- Why is this change being made? -->

This will add the ability to fully control the path of the created binary by `goreleaser`.  This comes in handy, when you create for example a (ZIP) archive, for distributing Terraform providers inside a CI/CD pipeline and you use `goreleaser` as a matrix build tool.

<!-- # Provide links to any relevant tickets, URLs or other resources -->

caarlos0 commented:
hey @tmeckel, thanks for the pr!

just want to understand: you need this to change the path of the binary inside the zip file, correct?

if so, if we had file mappings in the archive pipe, this won't be needed, correct?

tmeckel commented:
@caarlos0 yes one of the benefits of having the ability  to fully control the output path can be the fact you can simply archive the created artifacts without having to fiddle with some sort of "file mappings" and stuff. Either with the "archive pipe" you mentioned or with another (external) script. But aside that, further processing of the goreleaser artifacts is often much more easier when you can control where they are created, than arduously rearranging them afterwards. 

Even if `goreleaser` is mainly focussing on creating "releases", it's a really great matrix build tool compared to tools like gox or somethings. And even better, if you start using `goreleaser` simply for creating artifacts for multiple target platforms, you don't have to switch your tool set later, if you're required to create fully fledged releases.

Apropos archive pipeline and file mappings: according to the documentation of the archive pipeline (https://goreleaser.com/customization/archive/) I don't have any clue how I could rearrange (manipulate) the files paths inside the created archive. Any guidance (help)?

caarlos0 commented:
> @caarlos0 yes one of the benefits of having the ability to fully control the output path can be the fact you can simply archive the created artifacts without having to fiddle with some sort of "file mappings" and stuff. Either with the "archive pipe" you mentioned or with another (external) script. But aside that, further processing of the goreleaser artifacts is often much more easier when you can control where they are created, than arduously rearranging them afterwards.
> 
> Even if `goreleaser` is mainly focussing on creating "releases", it's a really great matrix build tool compared to tools like gox or somethings. And even better, if you start using `goreleaser` simply for creating artifacts for multiple target platforms, you don't have to switch your tool set later, if you're required to create fully fledged releases.

hmm, yeah, assuming the usage of `goreleaser build` only, it makes sense indeed.



> Apropos archive pipeline and file mappings: according to the documentation of the archive pipeline ([goreleaser.com/customization/archive](https://goreleaser.com/customization/archive/)) I don't have any clue how I could rearrange (manipulate) the files paths inside the created archive. Any guidance (help)?

its not yet possible, see #1638

caarlos0 commented:
One possible issue I see with this: the user could define the "same" path for all binaries, which means builds would override each other 🤔 

not sure if we need to fix/prevent it, but maybe document? 

Also, maybe we could instead make `binary` templateable? 🤔

tmeckel commented:
> 
> 
> One possible issue I see with this: the user could define the "same" path for all binaries, which means builds would override each other 🤔
> 
> not sure if we need to fix/prevent it, but maybe document?
> 
when this could be the case, then this should definitely be documented. This could cause headaches 👎🏼 

> Also, maybe we could instead make `binary` templateable? 🤔

Well, if this could be done without creating a breaking change it would be a much more general solution than having a new configuration property :+1

caarlos0 commented:
> Well, if this could be done without creating a breaking change it would be a much more general solution than having a new configuration property :+1

afaik it could yeah... once upon a time it couldn't because docker builds would need to specify which binaries to copy, but that's no longer the case...

tmeckel commented:
> 
> 
> > Well, if this could be done without creating a breaking change it would be a much more general solution than having a new configuration property :+1
> 
> afaik it could yeah... once upon a time it couldn't because docker builds would need to specify which binaries to copy, but that's no longer the case...

So what would you proposal here. Should I cancel my PR? Or should I add the functionality for templating the `binary` setting to this PR. If so what would you propose how this should be implemented.

caarlos0 commented:
I was digging into it... so, binary is templateable already... the issue is that it still puts the binaries in a subfolder in dist, e.g.:

```yaml
binary: "{{.Os}}/{{.Arch}}/{{.ProjectName}}_{{.Os}}_{{.Arch}}"
```

becomes

```sh
./dist/goreleaser_linux_arm64/linux/arm64/goreleaser_linux_arm64
```

so, I'm wondering if this is just a matter of lack of documentation or if you need the binaries to not be wrapped in `{{ .ProjectName }}_{{ .Os }}_{{ .Arch }}`?

caarlos0 commented:
FWIW, just tried a config like this here:

```yaml
builds:
  - env:
      - CGO_ENABLED=0
    goos:
      - linux
      - darwin
    goarch:
      - arm64
      - amd64
    binary: "{{.Os}}/{{.Arch}}/{{.ProjectName}}_{{.Os}}_{{.Arch}}"

```

and archives also look as what I think you would expect:

```sh
❯ tar tvf dist/foobar_v0.0.0-SNAPSHOT-7efc141_linux_amd64.tar.gz 
-rwxrwxr-x carlos/carlos 1355776 2021-06-02 02:08 linux/amd64/foobar_linux_amd64
```

tmeckel commented:
> 
> 
> I was digging into it... so, binary is templateable already... the issue is that it still puts the binaries in a subfolder in dist, e.g.:
> 
> ```yaml
> binary: "{{.Os}}/{{.Arch}}/{{.ProjectName}}_{{.Os}}_{{.Arch}}"
> ```
> 
> becomes
> 
> ```shell
> ./dist/goreleaser_linux_arm64/linux/arm64/goreleaser_linux_arm64
> ```
> 
> so, I'm wondering if this is just a matter of lack of documentation or if you need the binaries to not be wrapped in `{{ .ProjectName }}_{{ .Os }}_{{ .Arch }}`?

Exactly that's why I introduced the `output_path` property, because the current code in `build.go` uses a static pattern to build to output path:

![image](https://user-images.githubusercontent.com/14177833/120506031-d685d200-c3c5-11eb-959c-006cc5d459d5.png)

and because I'm fan of not introducing breaking changes (especially for a tool like `goreleaser` that is heavily used!) I changed the generation of  the output path as the following, where the old strategy is the default and is only overwritten when the `output_path` is set:

![image](https://user-images.githubusercontent.com/14177833/120506419-30869780-c3c6-11eb-8c20-5a8a18619720.png)

but I've to admit that if there would be a way to fully control the output path **without** introducing a new config setting this would be great! 👍🏼 

And about your test with the `binary` settings using a pattern `"{{.Os}}/{{.Arch}}/{{.ProjectName}}_{{.Os}}_{{.Arch}}"`:

This will still create those "default" folders inside the `dist` directory. The pattern is only effective for this sub directory.

I used the following pattern:

`binary: "microsoft/azuredevops/{{ .Version }}/{{ .Os }}_{{ .Arch }}/terraform-provider-azuredevops" `

which creates the following directories inside dist:

![image](https://user-images.githubusercontent.com/14177833/120507139-d508d980-c3c6-11eb-828d-8d69f9a0ba01.png)

But my approach creates the following structure, when the `output_path` is set like the following and the `binary` settings is set to a static string:

* `binary: "terraform-provider-azuredevops"`
* `output_path: "microsoft/azuredevops/{{ .Version }}/{{ .Os }}_{{ .Arch }}"`

![image](https://user-images.githubusercontent.com/14177833/120507470-21ecb000-c3c7-11eb-9719-db2b8c7b5850.png)

With this directory structure I'm able create a ZIP file, which contains *all* binary versions for the different target platforms in one single shot:

```shell
$ cd dist
$ zip -9rv microsoft-azuredevops-1.1.0.zip microsoft
```

This ZIP file then has the required directory structure that is expected the `terraform-bundle` command.

![image](https://user-images.githubusercontent.com/14177833/120508189-ce2e9680-c3c7-11eb-9b95-b5a6282bb727.png)

I hope I could clarify my use-case and motivation for the PR 😀

caarlos0 commented:
hmm got it...

will need to think for a while on a good way to introduce this without breaking changes...

I think we could assume most people only use the `binary` property to change the binary name instead of its full path, and for the people that do change it, I think they would expect it to change the full path (like you do)

so, far, 2 ideas:

## 1

- on Defaults, check if binary has a `/` or `\`, if not, prefix with `{{.ProjectName}}_{{.Os}}_{{.Arch}}{{.Arm}}{{.Mips}}` (or something like that)
- remove that sprintf

## 2

- add a new `wrap_dir` (or something with a better name) that defaults to true on build
- only if its true prepend the sprintf thing

2nd option seems to be more backwards-compatible... just need to find a good name for it :P

tmeckel commented:
# 3rd

Accept my PR and go with `output_path` 👍🏼  😀 


But either way I'm happy that I got my point across and that you grasped the use case. I'm fine with any solution here 😸

caarlos0 commented:
having two options in which you can specify the path looks a bit confusing IMHO...

tmeckel commented:
> 
> 
> having two options in which you can specify the path looks a bit confusing IMHO...

That's why I said it's kinda crazy... but if it comes to me I set the `binary` setting to a constant value (obviously because I wasn't aware of the fact that's possible to use a template string 😁 ) but you can consider `output_path` as a "base path" or something as I wrote in the "documentation".

But I agree _ease of use_ and _predictability_ are two very valid factors one should consider designing a tool or API.

tmeckel commented:
> 
> 2nd option seems to be more backwards-compatible... just need to find a good name for it :P

A good name for the option might be `architecture_dir` or `create_architecture_dir` because in the current implementation it's the path for the combination of  `.Os` and `.Arch` in a matrix build.

caarlos0 commented:
check https://github.com/goreleaser/goreleaser/pull/2280 plz :) @tmeckel
